moto/s3/responses.py

```diff
@@ -1313,21 +1315,21 @@ def _key_response(
             if bucket_permissions == PermissionResult.DENIED:
                 return 403, {}, ""
 
-            # If the request is not authorized, and not signed,
-            # that means that the action should be allowed for anonymous users
-            if not authorized_request and not signed_url:
-                # We already know that the bucket permissions do not explicitly deny this
-                # So bucket permissions are either not set, or do not explicitly allow
-                # Next check is to see if the ACL of the individual key allows this action
-                if bucket_permissions != PermissionResult.PERMITTED and (
-                    key.acl and not key.acl.public_read
-                ):
-                    return 403, {}, ""
-
-        elif signed_url and not authorized_request:
-            # coming in from requests.get(s3.generate_presigned_url())
-            if self._invalid_headers(request.url, dict(request.headers)):
-                return 403, {}, S3_INVALID_PRESIGNED_PARAMETERS
+            if key:
+                # If the request is not authorized, and not signed,
+                # that means that the action should be allowed for anonymous users
+                if not authorized_request and not signed_url:
+                    # We already know that the bucket permissions do not explicitly deny this
+                    # So bucket permissions are either not set, or do not explicitly allow
+                    # Next check is to see if the ACL of the individual key allows this action
+                    if bucket_permissions != PermissionResult.PERMITTED and (
+                        key.acl and not key.acl.public_read
+                    ):
+                        return 403, {}, ""
+            elif signed_url and not authorized_request:
```

Comment:
If we had rewritten the original if-statement, the full logic was:
`if not key and signed_url and not authorized_request`

If we indent the if-statement here, the logic changes to:
`if bucket and not key and signed_url and not authorized_request`

So in order to keep this logic the same, I think we should revert the indentation and change the logic to make it explicit:

```suggestion
        if not key and signed_url and not authorized_request:
```

Whether the bucket exists or not shouldn't matter here (I think...), so it shouldn't be inside the `if bucket`-statement.

(Unless I'm missing something? Feel free to push back if I've got something wrong in the logic!)

Reply:
I indented it mainly because `key` is `None` when the bucket doesn't exist according to this:
```
try:
    key = self.backend.get_object(bucket_name, key_name)
    bucket = self.backend.get_bucket(bucket_name)
except S3ClientError:
    key = bucket = None
```
so I thought it would be better to skip the condition when the bucket is missing. I can reverse the indentation to make the logic the same as it was before, the difference is negligible I think, but I may be missing something

Reply:
But, by not indenting it, wouldn't that put the possibility that it would return a 403 if the bucket doesn't exist?

Reply:
I've just tested this against AWS - they always throw a 403 if the headers are wrong/missing, regardless of whether the bucket exists or not. 

To be completely correct, we would have to move this entire if-statement to the top, and simplify it to: `if signed_url and not authorized_request and self._invalid_headers()`.
Only after that do we need to check if the bucket/key even exists.

But let's keep it as is for now - one improvement at the time :slightly_smiling_face: